src/access/RequiredPrivilege.cpp

```diff
@@ -0,0 +1,214 @@
+// Sorting the entries in this way allows a cluster to override specific endpoints
+// or fields with privilege, while still using wildcard to override all the
+// unspecified ones.
+struct PrivilegeOverride
+{
+    ClusterId mCluster;
+    EndpointId mEndpoint;
+    FieldId mField;
+    Privilege mPrivilege;
+
+    PrivilegeOverride() : mCluster(kInvalidClusterId), mEndpoint(kInvalidEndpointId), mField(kInvalidFieldId) {}
+
+    PrivilegeOverride(ClusterId cluster, EndpointId endpoint, FieldId field, Privilege privilege) :
+        mCluster(cluster), mEndpoint(endpoint), mField(field), mPrivilege(privilege)
+    {}
+
+    static_assert(sizeof(FieldId) >= sizeof(AttributeId), "FieldId must be able to hold AttributeId");
+    static_assert(sizeof(FieldId) >= sizeof(CommandId), "FieldId must be able to hold CommandId");
+    static_assert(sizeof(FieldId) >= sizeof(EventId), "FieldId must be able to hold EventId");
+};
+
+PrivilegeOverride
+    privilegeOverrideForReadAttribute[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_READ_ATTRIBUTE] = {
+        PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+    };
+
+PrivilegeOverride
+    privilegeOverrideForWriteAttribute[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_WRITE_ATTRIBUTE] = {
+        PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+    };
+
+PrivilegeOverride
+    privilegeOverrideForInvokeCommand[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_INVOKE_COMMAND] = {
+        PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+    };
+
+PrivilegeOverride privilegeOverrideForReadEvent[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_READ_EVENT] = {
+    PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+};
+
+enum class Operation
+{
+    kReadAttribute  = 0,
+    kWriteAttribute = 1,
+    kInvokeCommand  = 2,
+    kReadEvent      = 3
+};
+
+PrivilegeOverride * const privilegeOverride[] = { privilegeOverrideForReadAttribute, privilegeOverrideForWriteAttribute,
+                                                  privilegeOverrideForInvokeCommand, privilegeOverrideForReadEvent };
+
+const size_t maxPrivilegeOverride[] = { sizeof(privilegeOverrideForReadAttribute) / sizeof(privilegeOverrideForReadAttribute[0]),
+                                        sizeof(privilegeOverrideForWriteAttribute) / sizeof(privilegeOverrideForWriteAttribute[0]),
+                                        sizeof(privilegeOverrideForInvokeCommand) / sizeof(privilegeOverrideForInvokeCommand[0]),
+                                        sizeof(privilegeOverrideForReadEvent) / sizeof(privilegeOverrideForReadEvent[0]) };
+
+Privilege GetRequiredPrivilege(Operation operation, ClusterId cluster, EndpointId endpoint, FieldId field,
+                               Privilege defaultPrivilege)
```

Comment:
Do we need to set the defaultPrivilege for each consumer?  could we derive this value based on operation?

Reply:
It is derived based on operation, see where it comes from. (This is just a code-sharing thing.)

Reply:
My point is, if the defaultPrivilege is always same for a specific operation, then we don't have to have Privilege defaultPrivilege in the function parameter list.

Reply:
I see now what you're saying. Sure, I moved it into an array.